### PR TITLE
Add tracker support for pre-execution

### PR DESCRIPTION
**Pull request by yontyon**
Here we add support in pre execution for the tracker as well as adding tracker to the pre-execution tests

### CODE REVIEW

File: tests/apollo/test_skvbc_network_partitioning.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ class SkvbcNetworkPartitioningTest(unittest.TestCase):
 
     @with_trio
     @with_bft_network(start_replica_cmd)
-    @verify_linearizability
+    @verify_linearizability(pre_exec_enabled=False)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_network_partitioning.py:50]: How about simply:
`@verify_linearizability()`, in order to use the default value?

File: tests/apollo/util/skvbc_history_tracker.py [hunk 2]
```diff
@@ -836,17 +843,17 @@ async def get_last_block_id(self, client):
         msg = kvbc.SimpleKVBCProtocol.get_last_block_req()
         return kvbc.SimpleKVBCProtocol.parse_reply(await client.read(msg))
 
-    async def send_tracked_write(self, client, max_set_size):
+    async def send_tracked_write(self, client, max_set_size, long_exec=False):
         readset = self.readset(0, max_set_size)
         writeset = self.writeset(max_set_size)
         read_version = self.read_block_id()
-        msg = self.skvbc.write_req(readset, writeset, read_version)
+        msg = self.skvbc.write_req(readset, writeset, read_version, long_exec)
```
Comment (teoparvanov) [inline tests/apollo/util/skvbc_history_tracker.py:850]: How is this related to pre-execution?

File: tests/apollo/util/skvbc_history_tracker.py [hunk 3]
```diff
@@ -375,7 +377,12 @@ class SkvbcTracker:
     clusters with lots of blocks, but we may want to add it as an optional check
     in the future.
     """
-    def __init__(self, initial_kvpairs={}, skvbc=None, bft_network=None):
+    def __init__(self, initial_kvpairs={}, skvbc=None, bft_network=None, pre_exec=False):
+
+        # If this flag is set to True, it means that all the tracker requests will
+        # go through the pre_execution mechanism
+        self.pre_exec = pre_exec
```
Comment (teoparvanov) [inline tests/apollo/util/skvbc_history_tracker.py:384]: pre_execute_all ?

File: tests/apollo/test_skvbc_network_partitioning.py (hunk 1 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_network_partitioning.py]: done

File: tests/apollo/util/skvbc_history_tracker.py (hunk 2 above)
Comment (yontyon) [reply tests/apollo/util/skvbc_history_tracker.py:850]: We have one test in the pre execution tests that build a message with long_exec

File: tests/apollo/util/skvbc_history_tracker.py (hunk 3 above)
Comment (yontyon) [reply tests/apollo/util/skvbc_history_tracker.py]: done

File: tests/apollo/test_skvbc_preexecution.py [hunk 4]
```diff
@@ -94,49 +95,48 @@ async def run_concurrent_pre_execution_requests(self, skvbc, clients, num_of_req
...
-        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
 
         client = bft_network.random_client()
         client.config = client.config._replace(req_timeout_milli=LONG_REQ_TIMEOUT_MILLI, retry_timeout_milli=1000)
-        await self.send_single_write_with_pre_execution(skvbc, client, long_exec=True)
+        await tracker.send_tracked_write(client, 2, long_exec=True)
 
-        clients = bft_network.clients.values()
         with trio.move_on_after(seconds=1):
-            await self.send_indefinite_pre_execution_requests(skvbc, clients)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_preexecution.py:139]: Can we delete `send_indefinite_pre_execution_requests` ?

File: tests/apollo/test_skvbc_preexecution.py [hunk 5]
```diff
@@ -156,12 +157,8 @@ async def test_view_change(self, bft_network):
 
         clients = bft_network.clients.values()
         client = random.choice(list(clients))
-        key_before_vc = skvbc.random_key()
-        value_before_vc = skvbc.random_value()
-        write_set = [(key_before_vc, value_before_vc)]
 
-        await self.send_single_write_with_pre_execution_and_kv(skvbc, write_set, client)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_preexecution.py:163]: Same here, I guess send_single_write_with_pre_execution_and_kv() is now dead code?

File: tests/apollo/test_skvbc_preexecution.py (hunk 4 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_preexecution.py:139]: We can, but don't you think we should not remove this code to a case we will run a test without the tracker?

File: tests/apollo/test_skvbc_preexecution.py (hunk 5 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_preexecution.py:163]: same

File: tests/apollo/test_skvbc_preexecution.py (hunk 4 above)
Comment (teoparvanov) [reply tests/apollo/test_skvbc_preexecution.py:139]: agreed, thanks.
